### ISSUE TITLE
[TestRunner] tsconfig, babel, tsx, esm metabug

### ISSUE THREAD

**Issue by pavelfeldman** [microsoft/playwright#7121]
Module resolver issue extracted as https://github.com/microsoft/playwright/issues/7066.

https://github.com/microsoft/playwright/issues/7013
https://github.com/microsoft/playwright/issues/7037
https://github.com/microsoft/playwright/issues/7041
https://github.com/microsoft/playwright/issues/7042
https://github.com/microsoft/playwright/issues/7142
https://github.com/microsoft/playwright/issues/7148
https://github.com/microsoft/playwright/issues/7294
https://github.com/microsoft/playwright/issues/7855

**Comment by curiousercreative** [comment]
Is this a regression that may be temporarily worked around with an older version?

**Comment by mxschmitt** [comment]
This is a meta issue where we collect feedback to make @playwright/test compatible with custom TypeScript, Babel, and ESM. In what of these are you currently interested and what is your use-case?

**Comment by curiousercreative** [comment]
@mxschmitt I'm new to Playwright test running and simply trying to get an ESM test to run. I tried some other versions of `@playwright/test` but none seemed to allow for .js files (or .mjs files) to use ESM syntax. Using a .ts extension seems to appease it for now though...

**Comment by mxschmitt** [comment]
Yes, .mjs is not supported currently. We recommend TypeScript for now until we have added MJS support.

**Comment by zuzusik** [comment]
Doubling down on https://github.com/microsoft/playwright/issues/7066

We are considering to move from Mocha to Playwright's own test runner.

With Mocha we were able to use `ts-node` in combination with `tsconfig-paths` to get `tsconfig.json` `paths` feature to work. We _really_ wouldn't want to go back to "../../../blah" style of imports...

**Comment by bennypowers** [comment]
https://github.com/microsoft/playwright/issues/7452 is probably a dupe of this issue?

**Comment by cameronbraid** [comment]
My app uses module imports like
```
import { AffiliateConfig } from "src/util/affiliate";
```

but these don't work in playwright out of the box

in Jest config I use 

```
module.exports = {
  moduleNameMapper: {
    '^src/(.*)': '<rootDir>/src/$1',
  }
}
```

in tsconfig I have 
```
{
  "compilerOptions": {
    "paths": {
      "src/*" : ["src/*"]
    }
}
```

And in playwright config I managed to get it working with `app-module-path` :

```
import { PlaywrightTestConfig } from '@playwright/test';
import appModulePath from "app-module-path"

appModulePath.addPath(__dirname)

const config: PlaywrightTestConfig = {
  use: {
    
  },
  retries: 2
};
export default config;
```

**Comment by cyrilletuzi** [comment]
First, thanks for this tool!

I can't use Playwright in an Angular project (with a classic official Angular CLI configuration), because of legacy decorators:

> SyntaxError: /path/to/src/app/some.service.ts: Support for the experimental syntax 'decorators-legacy' isn't currently enabled (4:1):

(so with `@Injectable()` in `some.service.ts`)

> at parser (/path/to/node_modules/@babel/core/lib/parser/index.js:52:34)
>    at parser.next (<anonymous>) {
>  loc: Position { line: 4, column: 0 },
>  pos: 112,
>  missingPlugin: [ 'decorators-legacy', 'decorators' ],
>  code: 'BABEL_PARSE_ERROR'

And as I didn't find a way to override the babel config, I'm stuck.

More generally, it feels weird to be dependent on (complex) babel config files in a primary-TypeScript tool. It should at least do what's needed to respect the `tsconfig.json`, where `experimentalDecorators` is enabled.

**Comment by andreifloricel** [comment]
Hi, 
regarding #7066: is it planned to implement this (support of TS path aliases)?
I'm asking because this is a deal-breaker for us and before starting a big migration I need to decide between the playwright-test or playwright-jest runers.

**Comment by abierbaum** [comment]
https://github.com/microsoft/playwright/issues/7911 is the key pain point for me right now.  I expect that I will need more support for customizing tsconfig.json, but the lack of module path mapping is a deal breaker in my codebase and prevents me from using playwright test within my standard typescript project.

**Comment by bhanna1693** [comment]
> First, thanks for this tool!
> 
> I can't use Playwright in an Angular project (with a classic official Angular CLI configuration), because of legacy decorators:
> 
> > SyntaxError: /path/to/src/app/some.service.ts: Support for the experimental syntax 'decorators-legacy' isn't currently enabled (4:1):
> 
> (so with `@Injectable()` in `some.service.ts`)
> 
> > at parser (/path/to/node_modules/@babel/core/lib/parser/index.js:52:34)
> > at parser.next () {
> > loc: Position { line: 4, column: 0 },
> > pos: 112,
> > missingPlugin: [ 'decorators-legacy', 'decorators' ],
> > code: 'BABEL_PARSE_ERROR'
> 
> And as I didn't find a way to override the babel config, I'm stuck.
> 
> More generally, it feels weird to be dependent on (complex) babel config files in a primary-TypeScript tool. It should at least do what's needed to respect the `tsconfig.json`, where `experimentalDecorators` is enabled.

I have also been getting this issue! We use openapi generation for our interfaces and http services. The problem for us was we were trying to use interfaces in the openapi folder (which we configured to use TS path in the tsconfig file). For some reason, playwright was trying to read ALL the files (including the services) and hence why we were getting the error. Fixed this by adjusting imports to use exact relative paths.

**Comment by cyrilletuzi** [comment]
@bhanna1693 Thank you for your feedback. Unfortunately, except if if I'm missing something, your solution is okay if you just want to do e2e tests, but if you also want to use Playwright to do unit tests, then the service (and thus the decorator) has to be read.

**Comment by andreifloricel** [comment]
Hi @pavelfeldman @mxschmitt,
sorry to be nagging, but support for module path mapping (#7911) was asked multiple times by several people.
As I said, this is a deal-breaker when choosing between playwright-test or playwright-jest and it would be really helpful if we would know if this feature is planned (or not) in the near future.
Thanks

**Comment by zhex900** [comment]
please add module path mapping

**Comment by rhyek** [comment]
I have a monorepo with several TypeScript apps and I have avoided the use of Babel throughout. For testing I use `ts-jest` which uses my existing `tsc` and `tsconfig.json` under the hood. These are configurable as well. This has worked great for me so far since for path aliases I have been using `ts-patch` and the `typescript-transform-paths` plugin for normal app building, so ts-jest will use that same patched tsc program. I can maintain parity between runtime and testing TypeScript functionality quite easily.

I think playwright's test runner should do the same thing.

**Comment by thernstig** [comment]
I noticed that `.mjs` is supported now, which is nice. But many projects, including mine, using `"type": "module"`. But according to https://github.com/microsoft/playwright/issues/7452#issuecomment-874858986 this does not even seem planned to support.

We want to use `"type": "module"` for all our uses, and not require .mjs on any test files.

**Comment by actuallyrob** [comment]
I have created a Playwright configuration for my monorepo. For now I am only interested in running e2e tests with Playwright.
I have created two projects in my configuration for each app in my monorepo. Each project has its own testDir and testMatch (`**/*.e2e.spec.ts`) defined. When I run `npx playwright test list`, Playwright reports that no tests are found. I do however expect my tests that are defined in my test folder to be listed.

When I run `npx playwright test`, Playwright tries to parse .ts files that do not match my testMatch (and fails because of Aurelia decorators). Why does Playwright do this, and why is my testMatch not working?

**Comment by aslushnikov** [comment]
> When I run npx playwright test, Playwright tries to parse .ts files that do not match my testMatch (and fails because of Aurelia decorators). Why does Playwright do this, and why is my testMatch not working?

Could you please file this separately?

**Comment by andrewdavey** [comment]
My work-around to avoid all the "../../../foo" noise is to first set the `NODE_PATH` environment variable to be the test code directory. Then any imports like "foo" resolve correctly.

**Comment by Lonli-Lokli** [comment]
As an addition for paths support, decorators from tsconfig are also great feature which needs to be supported!

**Comment by DOrlov77** [comment]
Hello, any plans to fix this issue - the paths support?

**Comment by thernstig** [comment]
> I noticed that `.mjs` is supported now, which is nice. But many projects, including mine, using `"type": "module"`. But according to [#7452 (comment)](https://github.com/microsoft/playwright/issues/7452#issuecomment-874858986) this does not even seem planned to support.
> 
> We want to use `"type": "module"` for all our uses, and not require `.mjs` on any test files.

Should I file a separate issue for this? It's hard to track this meta-issue on what is planned and what is not.

**Comment by artemrudenko** [comment]
Hi, is there any plans to fix this? We are actively using decorators, paths and etc. Sure we can change paths, but what can we do with decorators? I believe that playwright should support all the range of options from tsconfig natively.

**Comment by Gojosrepo** [comment]
Just to add, it seems support for module path mapping (#7911) also fails for jsconfig as well. (basically the same but throwing it out there).

**Comment by awarecan** [comment]
[tsconfig-paths](https://www.npmjs.com/package/tsconfig-paths) works to me for #7911 

My tsconfig.json
```json
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es6",
    "outDir": "build",
    "baseUrl": ".",
    "paths": {
      "@my/common-libs": [
        "./modules/common-libs/src"
      ]
    }
  }
}
```

Add a `tsconfig-paths-bootstrap.js` to my project root folder
```javascript
const tsConfig = require("./tsconfig.json");
const tsConfigPaths = require("tsconfig-paths");

const baseUrl = "./build/"; // Either absolute or relative path. If relative it's resolved to current working directory.
tsConfigPaths.register({
    baseUrl,
    paths: tsConfig.compilerOptions.paths
});
```

Then call playwright likes
```sh
node -r ./tsconfig-paths-bootstrap.js ./node_modules/.bin/playwright test ...
```

**Comment by DOrlov77** [comment]
@awarecan Nice hack! But when tests are runing by CI systems you are not always may/want to deal with such things.

**Comment by irharrier2** [comment]
I hope that I haven't miss out anything before writing this comment but is there any update on this issue? We are considering using Playwright test runner but this is a deal-breaker for us.

**Comment by cameronbraid** [comment]
Another alias solution for yarn users  https://yarnpkg.com/features/protocols#why-is-the-link-protocol-recommended-over-aliases-for-path-mapping

**Comment by corysimmons** [comment]
@cameronbraid Pretty slick but in my specific case we're using `@/src/` as the alias.

...it's always something... Everything in the JS ecosystem is so brittle and there are so many different implementations of things. I spend so much more time trying to get tooling configured than actually making apps that provide any sort of value to humans. I'm about to just go to vanilla JS and never touch Github or NPM again. :(

**Comment by stabback** [comment]
I came across https://github.com/microsoft/playwright/issues/7042, which has been folded into this issue.

My use case - I have a monorepo with multiple packages.  In one test, I'm attempting to import some data from one package to help build the test.  That package happens to import from a third package.  That third package also exports `.tsx` files.  Playwright breaks, because it cannot handle `.tsx` files.

**Comment by OriAmir** [comment]
I have react app + e2e folder for the playwright testing.
I read all that thread and I'm still do not understand how I could make playwright to recognize absolute paths? I understand its not support the `tsconfig` file for now, but there is any workaround to make it work?

**Comment by pavelfeldman** [comment]
I'd like to give an update on this issue since there has been some development, which is available in our nightly. It is also about to hit 1.18 RC later this week. I intend to close this metabug and open us to new bugs that would be targeting missing specifics. Please don't hesitate to file those.

1. Playwright Test will now respect `tsconfig.json`'s `baseUrl` and `paths`, so you can use aliases
1. There is a new environment variable `PW_EXPERIMENTAL_TS_ESM` that allows importing ESM modules in your TS code, without the need for the compile step. Don't forget the `.js` suffix when you are importing your esm modules. Run your tests as follows:

```bash
npm i --save-dev @playwright/test@1.18.0-alpha-jan-10-2022
PW_EXPERIMENTAL_TS_ESM=1 npx playwright test
```

Let us know whether it worked for you in a new bug with your expectations and repro steps, link this issue for context please!

**Comment by chaitraAshu** [comment]
> > First, thanks for this tool!
> > I can't use Playwright in an Angular project (with a classic official Angular CLI configuration), because of legacy decorators:
> > > SyntaxError: /path/to/src/app/some.service.ts: Support for the experimental syntax 'decorators-legacy' isn't currently enabled (4:1):
> > 
> > 
> > (so with `@Injectable()` in `some.service.ts`)
> > > at parser (/path/to/node_modules/@babel/core/lib/parser/index.js:52:34)
> > > at parser.next () {
> > > loc: Position { line: 4, column: 0 },
> > > pos: 112,
> > > missingPlugin: [ 'decorators-legacy', 'decorators' ],
> > > code: 'BABEL_PARSE_ERROR'
> > 
> > 
> > And as I didn't find a way to override the babel config, I'm stuck.
> > More generally, it feels weird to be dependent on (complex) babel config files in a primary-TypeScript tool. It should at least do what's needed to respect the `tsconfig.json`, where `experimentalDecorators` is enabled.
> 
> I have also been getting this issue! We use openapi generation for our interfaces and http services. The problem for us was we were trying to use interfaces in the openapi folder (which we configured to use TS path in the tsconfig file). For some reason, playwright was trying to read ALL the files (including the services) and hence why we were getting the error. Fixed this by adjusting imports to use exact relative paths (not the ideal solution but it works).

Hello,
I'm facing the same issue while using services in playwright.  . I'm new to scripting language. If you Can explain how did you resolve it.   thanks in advance.
below is the error:
Support for the experimental syntax 'decorators-legacy' isn't currently enabled

**Comment by mxschmitt** [comment]
Please file new feature requests for it or create separate bug reports if you encounter issues.
